Userland/Libraries/LibCore/AnonymousBufferWindows.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2021, Andreas Kling <andreas@ladybird.org>
+ * Copyright (c) 2024, stasoid <stasoid@yahoo.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/AnonymousBuffer.h>
+#include <windows.h>
+
+namespace Core {
+
+AnonymousBufferImpl::AnonymousBufferImpl(AnonymousBufferId id, void* map_handle, size_t size, void* data)
+    : m_id(id)
+    , m_map_handle(map_handle)
+    , m_size(size)
+    , m_data(data)
+{
+}
+
+AnonymousBufferImpl::~AnonymousBufferImpl()
+{
+    if (m_data)
+        UnmapViewOfFile(m_data);
+
+    if (m_map_handle)
+        CloseHandle(m_map_handle);
+}
+
+ErrorOr<NonnullRefPtr<AnonymousBufferImpl>> AnonymousBufferImpl::create(size_t size)
+{
+    static int next_id_tag;
+    ByteString name = ByteString::formatted("AB-{}-{}", GetCurrentProcessId(), next_id_tag++);
```

Comment:
Why do we need to name these? 

Per the docs https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga:

> [in, optional] lpName

> The name of the file mapping object.

> If this parameter matches the name of an existing mapping object, the function requests access to the object with the protection that flProtect specifies.

> If this parameter is NULL, the file mapping object is created without a name.

We shouldn't need to keep the name around. The only thing we should be pushing over IPC is a handle duplicated via `DuplicateHandle`, assuming those can be transferred over windows IPC mechanisms?

> Multiple processes can share a view of the same file by either using a single shared file mapping object or creating separate file mapping objects backed by the same file. A single file mapping object can be shared by multiple processes through inheriting the handle at process creation, duplicating the handle, or opening the file mapping object by name. For more information, see the [CreateProcess](https://learn.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa), [DuplicateHandle](https://learn.microsoft.com/en-us/windows/desktop/api/handleapi/nf-handleapi-duplicatehandle) and [OpenFileMapping](https://learn.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-openfilemappinga) functions.

I would rather that we keep the "anonymous" behavior of Anonymous buffers as much as possible. It shouldn't be possible for other random processes on the system running as the current user to intercept and read our mapped file data by guessing the names.

Reply:
Because using names is the only way to keep current architecture. `DuplicateHandle` requres target process pid. So IPC::Encoder would require to know target process pid.

I think we should worry only about inadvertent clash of names with other processes. If someone is malicious and tries to guess names they might also just read/write process memory directly with ReadProcessMemory.

Reply:
Does it not work to pass DUPLICATE_CLOSE_SOURCE in the options and null for hTargetProcessHandle? If we are ok with making a bunch of extra kernel objects, we could even first duplicate the handle passing ourselves, then duplicate the initial handle with "close source" to get an anonymous handle that can be passed to anyone. Assuming the API works the way I think it does, anyway.

We have a similar constraint for mach ports that we need the parent process to receive a send port right from the child in order to send it data over IPC. I recently added a new abstraction to LibIPC called "transport", which should help abstract away different ipc mechanisms between OS platforms.

Reply:
No, this doesn't work:
>[out] lpTargetHandle
...
This parameter is ignored if hTargetProcessHandle is NULL.

Reply:
I am currently working on EventLoop, and it has similar situation. When choosing how to implement Unix unnamed pipes and Unix domain sockets I chose Windows named pipes over sockets because when IPC::encode()ing a socket handle we need to know target process id for WSADuplicateSocket. (I don't use Windows unnamed pipes because they are not waitable)

I encourage you not to be very picky about implementation details at this point. I think when we have no functionality at all anything that works will do. If this architecture is deemed inappropriate it can always be upgraded later.

Reply:
Updated to use handle instead of name.